compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +442,206 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+static void appendDynamicDims(OpBuilder &b, Location loc,
+                              SmallVector<Value> &argumentDims, Value tensor) {
+  auto tensorType = tensor.getType().cast<RankedTensorType>();
+
+  // Fast-path for if the value comes from ops that support our dynamic
+  // shape interfaces. Otherwise we have to insert tensor.dim ops.
+  auto availableDims = IREE::Util::findDynamicDims(tensor);
+  if (availableDims.hasValue()) {
+    argumentDims.append(availableDims->begin(), availableDims->end());
+    assert(tensorType.getNumDynamicDims() == availableDims->size() &&
+           "not enough dynamic dims found");
+    return;
+  }
+
+  for (auto dim : llvm::enumerate(tensorType.getShape())) {
+    if (dim.value() != ShapedType::kDynamicSize) continue;
+    argumentDims.push_back(
+        b.createOrFold<tensor::DimOp>(loc, tensor, dim.index()));
+  }
+}
+
+/// Follow the reverse SSA use-def chain of the given value (always taking the
+/// tied operand) and return the first value outside of `regionOp`.
+static Optional<Value> findFirstTiedValueOutsideOfRegionOp(
+    Flow::DispatchRegionOp regionOp, Value value) {
+  // Check if `v` is defined outside of `regionOp`.
+  auto isOutside = [&](Value v) {
+    if (v.isa<OpResult>()) return !regionOp->isAncestor(v.getDefiningOp());
+    assert(v.isa<BlockArgument>() && "expected bbArg");
+    // DispatchRegionOp does not have block arguments.
+    return true;
+  };
+
+  while (!isOutside(value)) {
+    auto tiedOpInterface = value.getDefiningOp<IREE::Util::TiedOpInterface>();
+    if (!tiedOpInterface)
+      // Reached an op that does not implement the interface.
+      return llvm::None;
+    value = tiedOpInterface.getTiedResultOperand(value);
+    if (!value)
+      // Nothing is tied here.
+      return llvm::None;
+  }
+
+  return value;
+}
+
+/// Rewrite the DispatchRegionOp into a DispatchWorkgroupsOp. The
+/// DispatchRegionOp is not isolated from above and may capture any SSA value
+/// that is in scope. The generated DispatchWorkgroupsOp captures all SSA values
+/// explicitly and makes them available inside the region via block arguments.
+static FailureOr<Flow::DispatchWorkgroupsOp>
+rewriteFlowDispatchRegionToFlowDispatchWorkgroups(
+    Flow::DispatchRegionOp regionOp, RewriterBase &rewriter) {
+  // Only ops with a single block are supported.
+  Region &region = regionOp.getBody();
+  if (!region.hasOneBlock()) return failure();
+  Block &body = region.front();
+  auto terminator = cast<Flow::ReturnOp>(body.getTerminator());
+  unsigned numResults = terminator->getNumOperands();
+
+  // Prepare rewriter.
+  OpBuilder::InsertionGuard guard(rewriter);
+  Location loc = regionOp.getLoc();
+  rewriter.setInsertionPoint(regionOp);
+
+  // Compute arguments of the dispatch region.
+  llvm::SetVector<Value> argumentsSet;
+  mlir::getUsedValuesDefinedAbove(region, argumentsSet);
+  // Unranked tensors are not supported.
+  if (llvm::any_of(argumentsSet, [](Value v) {
```

Comment:
We dont need this here. I think this is checked much earlier (or should by in any case.)

Reply:
made it an assertion